flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -726,10 +726,127 @@ class ExprLowering {
     return builder.create<fir::CoordinateOp>(getLoc(), ty, base, args);
   }
 
+  mlir::Value gen(const Fortran::lower::SymIndex &si,
+                  const Fortran::evaluate::ArrayRef &aref) {
+    auto loc = getLoc();
+    auto addr = si.getAddr();
+    auto arrTy = fir::dyn_cast_ptrEleTy(addr.getType());
+    auto eleTy = arrTy.cast<fir::SequenceType>().getEleTy();
+    auto refTy = fir::ReferenceType::get(eleTy);
+    auto base = builder.create<fir::ConvertOp>(loc, refTy, addr);
+    auto idxTy = builder.getIndexType();
+    auto one = builder.createIntegerConstant(idxTy, 1);
+    auto zero = builder.createIntegerConstant(idxTy, 0);
+    return std::visit(
+        Fortran::common::visitors{
+            [&](const Fortran::lower::SymIndex::Shaped &arr) {
+              mlir::Value delta = one;
+              mlir::Value total = zero;
+              assert(arr.shape.size() == aref.subscript().size());
+              for (const auto &pair : llvm::zip(arr.shape, aref.subscript())) {
+                auto val = builder.create<fir::ConvertOp>(loc, idxTy,genval(std::get<1>(pair)));
+                auto diff = builder.create<mlir::SubIOp>(loc, val, one);
+                auto prod = builder.create<mlir::MulIOp>(loc, delta, diff);
+                total = builder.create<mlir::AddIOp>(loc, prod, total);
+                delta =
+                    builder.create<mlir::MulIOp>(loc, delta, std::get<0>(pair));
+              }
+              return builder
+                  .create<fir::CoordinateOp>(loc, refTy, base,
+                                             llvm::ArrayRef<mlir::Value>{total})
+                  .getResult();
+            },
+            [&](const Fortran::lower::SymIndex::FullDim &arr) {
+              mlir::Value delta = one;
+              mlir::Value total = zero;
+              auto subct = aref.subscript().size();
+              auto asiter = arr.shape.begin();
+              auto subiter = aref.subscript().begin();
+              for (auto asct = arr.shape.size(); asct;
+                   --asct, ++asiter, ++subiter, --subct) {
+                auto val = builder.create<fir::ConvertOp>(loc, idxTy,genval(*subiter));
+                auto lb = builder.create<fir::ConvertOp>(loc, idxTy, std::get<0>(*asiter));
+                auto diff = builder.create<mlir::SubIOp>(loc, val, lb);
+                auto prod = builder.create<mlir::MulIOp>(loc, delta, diff);
+                total = builder.create<mlir::AddIOp>(loc, prod, total);
+                delta = builder.create<mlir::MulIOp>(loc, delta,
+                                                     std::get<1>(*asiter));
+              }
+              if (subct) {
+                assert(subct == 1);
+                auto val = builder.create<fir::ConvertOp>(loc, idxTy,genval(*subiter));
+                auto diff = builder.create<mlir::SubIOp>(loc, val, one);
```

Comment:
Thanks. I'll have to test that and fix the bugs.

Reply:
I did a spot check and this case seems to work. (I think it was working already, so that's good.) For an input of
```Fortran
  real :: a(10:94, 4:*)
  do j = 10,94
    do i = 4, 40
      ... a(i,j) ...
```
lowering is generating
```mlir
   %13 = subi %12, %c10_i64 : i64
   %16 = subi %15, %c4_i64 : i64
   %17 = fir.coordinate_of %0, %13, %16 : (!fir.ref<!fir.array<85x?xf32>>, i64, i64) -> !fir.ref<f32>
```
So both lower bound adjustments look to be accounted for.